Cleanup of various virtual destructors

MaxKellermann opened the pull request:
Many virtual destructors are never used as such (while adding overhead), and this PR removes lots of them (plus related cleanups).

MaxKellermann committed:
PartyModeManager: remove useless virtual destructor

MaxKellermann committed:
addons/AddonEvents: remove virtual destructor

No derived class instance is ever allocated dynamically.

MaxKellermann committed:
LangInfo: remove virtual CRegion destructor

MaxKellermann committed:
LangInfo: remove custom CRegion copy constructor

The implicit copy constructor is just as good, and not so fragile.

MaxKellermann committed:
utils/EGLUtils: make CEGLContextUtils destructor non-virtual

MaxKellermann committed:
commons/Buffer: remove useless explicit destructor

MaxKellermann committed:
utils/TextSearch: remove useless virtual destructor and make class final

MaxKellermann committed:
utils/HttpRangeUtils: remove virtual destructor and make class final

MaxKellermann committed:
utils/URIUtils: remove unimplemented constructor/destructor prototypes

MaxKellermann committed:
utils/BitstreamStats: remove virtual destructor and make class final

MaxKellermann committed:
utils/ActorProtocol: make ~Protocol() non-virtual

There is no dynamic `delete`, and thus this class needs no vtable.

MaxKellermann committed:
utils/JobManager: make CJobManager final and remove the virtual destructor

MaxKellermann committed:
utils/IRssObserver: remove virtual destructor

MaxKellermann committed:
utils/IArchivable: remove virtual destructor

MaxKellermann committed:
utils/ISortable: remove virtual destructor

MaxKellermann committed:
utils/IXmlDeserializable: drop useless interface

There is no IXmlDeserializable::Deserialize() call on a possibly
polymorphed object, and therefore the whole interface with virtual
methods is pointless.  This patch removes overhead for vtable
initialization and for indirect calls.

MaxKellermann committed:
utils/BooleanLogic: drop useless CBooleanLogicOperation destructor

The C++ compiler will automatically generate code to clear those
`std::vector`s.

MaxKellermann committed:
utils/BooleanLogic: drop useless constructor, hide destructor

MaxKellermann committed:
music/Song: make class final

MaxKellermann committed:
music/Song: remove useless explicit destructor

MaxKellermann committed:
music/tags/MusicInfoTag: make class final

MaxKellermann committed:
music/tags/MusicInfoTag: remove useless explicit destructor

MaxKellermann committed:
utils/StreamDetails: make leaf classes final

MaxKellermann committed:
utils/StreamDetails: declare virtual destructor in base class

Right now, the virtual destructor from ISerializable is inherited, but
once this goes away, this library will fail.

MaxKellermann committed:
pvr/EpgInfoTag: make class final

MaxKellermann committed:
pvr/PVRRecording: make class final

MaxKellermann committed:
pvr/PVRTimerInfoTag: make class final

MaxKellermann committed:
pvr/EpgInfoTag: remove useless explicit destructor

MaxKellermann committed:
pvr/PVRRecording: remove useless explicit destructor

MaxKellermann committed:
pvr/PVRTimerInfoTag: remove useless explicit destructor

MaxKellermann committed:
pvr/PVRRadioRDSInfoTag: make class final

MaxKellermann committed:
pvr/PVRRadioRDSInfoTag: remove useless explicit destructor

MaxKellermann commented:
There are compiler warnings:
```
delete called on non-final 'CVideoInfoTag' that has virtual functions but non-virtual destructor
```
I'll add missing the `final` keywords. Stand by.

pkerling commented:
All classes that are derived from should have either a virtual destructor or a protected non-virtual destructor. (cf https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#c35-a-base-class-destructor-should-be-either-public-and-virtual-or-protected-and-nonvirtual)

MaxKellermann commented:
Not exactly true - this CppCoreGuidelines article isn't phrased properly. The virtual destructor is only needed if `delete` is called on a base class (or as in the example: `std::unique_ptr` is used with a base class, which is technically the same). If only leaf (`final`) classes are ever deleted, the virtual destructor is useless.
But the compiler warnings are about leaf classes which are not declared `final`.

MaxKellermann commented:
I was wrong in this case - `CVideoInfoTag` is indeed not a leaf class, and there are other classes derived from it. So there needs to be a virtual destructor after all! There wasn't one before - it was implicitly declared by deriving from `IArchivable`, `ISerializable` and `ISortable` which used to have one. So this previously worked only by accident, and I removed that accident.....

MaxKellermann committed:
video/VideoInfoTag: add virtual destructor

`CFileItem::~CFileItem()` invokes `delete` on `CVideoInfoTag`
pointers, thus `CVideoInfoTag` needs a virtual destructor.

MaxKellermann committed:
picture/PictureInfoTag: add virtual destructor

`CFileItem::~CFileItem()` invokes `delete` on `CPictureInfoTag`
pointers, thus `CPictureInfoTag` needs a virtual destructor.

MaxKellermann commented:
Rebased and added explicit (defaulted) virtual destructors to `CVideoInfoTag` and `CPictureInfoTag`.

pkerling commented:
> Not exactly true - this CppCoreGuidelines article isn't phrased properly. The virtual destructor is only needed if delete is called on a base class

I was thinking about the cases where you removed the virtual destructor, for example `AddonEvents`. IMO you should make the destructor non-virtual and protected in order to prevent accidental deletion via the base class.
Actually I don't really get why `AddonEvents` even is structured like that and why a simple `enum` wouldn't have sufficed for the event type. But out of scope for this PR...

The same is true for all the interface classes you removed the virtual destructors from. You should definitely add a protected non-virtual destructor to prevent misuse.

> So this previously worked only by accident, and I removed that accident.....

The interfaces having virtual destructors was not an accident :-) But it's of course better to be explicit here.

MaxKellermann committed:
utils/BooleanLogic: declare virtual destructors

These classes currently inherit virtual destructors from
IXmlDeserializable, but this is fragile, so let's make this explicit.

MaxKellermann committed:
utils/BooleanLogic: make methods non-virtual

Nobody overrides these.

MaxKellermann commented:
Agree, protected+defaulted destructors are a good safety net in some cases.
I rebased and added a few of those.
The previous Jenkins failures are due to a "new" compiler warning which is not so new:
```/home/jenkins/workspace/FreeBSD/xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:593:19: warning: unused variable 'now' [-Wunused-variable]```
I have a patch for this, and I'll submit it in a separate PR.

MaxKellermann commented:
Sigh, had to amend one commit again because one `final` was invalid (occurred only on Android due to the very strange DRM code).

pkerling commented on xbmc/utils/BooleanLogic.h:
```diff
@@ -86,16 +86,13 @@ class CBooleanLogicOperation : public IXmlDeserializable
 };
 
-class CBooleanLogic : public IXmlDeserializable
+class CBooleanLogic
```
Why does this not have any destructor?

MaxKellermann replied:
You're right, this one should have a `protected` defaulted destructor.

pkerling commented on xbmc/utils/URIUtils.h:
```diff
@@ -28,9 +28,6 @@ class CURL;
 class URIUtils
```
Missing `final`?

MaxKellermann replied:
Hm, better convert to `namespace`. This isn't a class after all - it just abuses the `class` keyword to create a namespaceo.

pkerling replied:
+1

`StringUtils` is similar, but you don't have to change either one in this PR.

MaxKellermann commented:
Rebased & `protected` destructor added to `CBooleanLogic`.

MaxKellermann commented:
Jenkins failed:
```
/home/jenkins/workspace/LINUX-64-Wayland/xbmc/utils/ActorProtocol.h: In constructor ‘KODI::WINDOWING::WAYLAND::CWinSystemWayland::CWinSystemWayland()’:
/home/jenkins/workspace/LINUX-64-Wayland/xbmc/utils/ActorProtocol.h:88:3: error: ‘Actor::Protocol::~Protocol()’ is protected
```
So this is where the `protected` destructor approach fails: when somebody uses the base class.
Kodi's approach of deriving from classes for no apparent reasons (in this case `Actor::Protocol` doesn't have any virtual methods), instead of embedding an instance, is questionable. Maybe I should work towards making those "base" classes like `Actor::Protocol` final, so nobody submits code which derives from them...

MaxKellermann commented:
Reverted `Actor::Protocol::~Protocol()` to `public`. Let's see if this build succeeds.

pkerling commented:
> So this is where the protected destructor approach fails: when somebody uses the base class.

Yes, that approach is only for classes that are used exclusively to inherit from. Here the only place where it is instantiated directly is in Wayland, so maybe I should change it there.

> in this case Actor::Protocol doesn't have any virtual methods

So? Does that automatically make derivation meaningless?

MaxKellermann commented:
> So? Does that automatically make derivation meaningless?

No, but derivation is only useful if the derived class is not only technically derived, but also derived by its meaning; if it adds something to the base class.

The design of `Actor::Protocol` is hard to grasp, because it's completely undocumented. There are exactly two code comments, and they explain the obvious, and obvious comments are disposable.
From the name: an instance of this class describes one protocol. But that's wrong. This is not a protocol; this is a generic inter-thread message queue, and a message is an opaque buffer. The protocol is defined by the caller.

So if this is really just a message queue, why would anybody derive from it? Maybe to add functionality. But none of the derived classes add queue-specific functionality. All they do is serve as a namespace containing protocol-specific definitions (opaque to the queue itself!). No derived class accesses these definitions; they are accessed solely from the outside, not even by an instance, but only by the namespace name:
```
          m_controlPort.SendInMessage(COutputControlProtocol::STATS);
```
It does not matter if `m_controlPort` is a `COutputControlProtocol` or a basic `Actor::Protocol` - there isn't even any wrapper ensuring that a `COutputControlProtocol` command can't be sent over a `CMixerControlProtocol` instance. The way this is used here is clumsy at best.
What would have been *somewhat* helpful would have been this choice of syntax:
```
          m_controlPort.SendInMessage(m_controlPort.STATS);
```
This would make it more obvious to the (human) reader, but still the compiler wouldn't care.
So no, in this case, deriving from `Actor::Protocol` doesn't make sense. One could make a wrapper template to ensure type-safety, to validate all protocol commands at compile-time, and that wrapper would embed that `Actor::Protocol` instance instead of deriving from it, providing inline methods wrapping `Actor::Protocol` methods in a type-safe way. Still nobody would derive from the latter.

(This is just my opinion, not the absolute truth, of course - it's a matter of taste and preferred coding style. There are reasons for/against one or another.)

MaxKellermann commented:
btw. Jenkins failures again not related to this PR.

pkerling commented:
ping @FernetMenta, Actor is your code so you might want to take a look

@MaxKellermann You probably won't mind if I merge this as-is for the time being?

MaxKellermann commented:
Of course I don't mind. This PR appears to be ready for merging (no known regression, just a false positive about an older compiler warning), and other design problems are mostly unrelated to this PR ...

Rechi commented:
> btw. Jenkins failures again not related to this PR.

They are, because for whatever reason the unused warning wasn't reported before.

> You probably won't mind if I merge this as-is for the time being?

Not as long as jenkins reports success.

MaxKellermann commented:
@Rechi I'll submit a PR for this warning. I have the patch already. No idea why it is regarded new suddenly - it's 2 years old.

MaxKellermann commented:
Oh you fixed it already.

pkerling commented:
@Rechi Sorry, was a bit too quick!

But I double-checked that the warning was true already before the PR. So for the record: All failed builds were built correctly and only failed in Jenkins due to one warning marked as new that clang did not flag before for whatever reason (even though it was already true before).
